### PR TITLE
syncpoint (ticdc): set downstream tidb_enable_external_ts_read = off

### LINKED ISSUE #7854: If tidb_enable_external_ts_read is set to on, Ticdc has no permission to synchronize data
### What did you do?

set global tidb_enable_external_ts_read=on;

### What did you expect to see?

Succeeded in synchronizing data.

### What did you see instead?

"code": "CDC:ErrMySQLTxnError",
        "message": "[CDC:ErrMySQLTxnError]MySQL txn error: Error 1105: only support read-only statement during read-only staleness transactions"

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
(paste TiDB cluster version here)
```

Upstream TiKV version (execute `tikv-server --version`):

```console
(paste TiKV version here)
```

TiCDC version (execute `cdc version`):
V6.4.0
```console
(paste TiCDC version here)
```

### PR DESCRIPTION
**Pull request by asddongmen**
### What problem does this PR solve?

Issue Number: close #7854 

#### Tests 

 - Integration test
 - Manual test (add detailed scripts or steps below)

### Release note 

```release-note
Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

If you don't think this PR needs a release note then fill it with `None`.
```

### CODE REVIEW

**Review by nongfushanquan — APPROVED** [review]

**Review by sdojjy — APPROVED** [review]

File: cdc/sink/mysql/mysql_syncpoint_store.go [hunk 1]
```diff
@@ -209,6 +210,19 @@ func (s *mysqlSyncPointStore) SinkSyncPoint(ctx context.Context,
 		log.Error("sync table: begin Tx fail", zap.Error(err))
 		return cerror.WrapError(cerror.ErrMySQLTxnError, err)
 	}
+
+	// we set `tidb_enable_external_ts_read` = off for each txn,
+	// otherwise TiCDC may not be able to write data to downstream TiDB when
+	// the downstream tidb_enable_external_ts_read = on.
+	if err = s.setTidbEnableExternalTsReadOff(ctx, tx); err != nil {
```
Comment (YangKeao) [inline cdc/sink/mysql/mysql_syncpoint_store.go:217]: The transaction is created before setting the variable, so it's not affected by this modification. We should set the variable before starting the transaction.

However, it's a little hard to do this in go mysql driver (here is a related discussing: https://github.com/go-sql-driver/mysql/issues/208). I'll try to find some way to solve this problem :thinking: .

Comment (YangKeao) [reply cdc/sink/mysql/mysql_syncpoint_store.go:217]: Maybe I could manually get a connection, and use this connection to setup the transaction and variables.

Comment (YangKeao) [reply cdc/sink/mysql/mysql_syncpoint_store.go:217]: I tried `s.db.Conn` and use this connection to set variables and start transaction, but it still doesn't work. I don't know why :facepalm:. Need some investigation.

**Comment by YangKeao** [comment]
Apply the following patch on this PR would make it work:

```
diff --git a/cdc/sink/mysql/mysql_params.go b/cdc/sink/mysql/mysql_params.go
index 0de0656f7..d2e4f890c 100644
--- a/cdc/sink/mysql/mysql_params.go
+++ b/cdc/sink/mysql/mysql_params.go
@@ -314,6 +314,16 @@ func generateDSNByParams(
        if tidbPlacementMode != "" {
                dsnCfg.Params["tidb_placement_mode"] = fmt.Sprintf(`"%s"`, tidbPlacementMode)
        }
+
+       tidbEnableExternalTSRead, err := checkTiDBVariable(ctx, testDB, "tidb_enable_external_ts_read", "OFF")
+       if err != nil {
+               return "", err
+       }
+       if tidbEnableExternalTSRead != "" {
+               // set the `tidb_enable_external_ts_read` to `OFF`, so cdc could write to the sink
+               dsnCfg.Params["tidb_enable_external_ts_read"] = fmt.Sprintf(`"%s"`, tidbEnableExternalTSRead)
+       }
+
        dsnClone := dsnCfg.Clone()
        dsnClone.Passwd = "******"
        log.Info("sink uri is configured", zap.String("dsn", dsnClone.FormatDSN()))
diff --git a/cdc/sink/mysql/mysql_syncpoint_store.go b/cdc/sink/mysql/mysql_syncpoint_store.go
index 3c1d25a5e..2c7657753 100644
--- a/cdc/sink/mysql/mysql_syncpoint_store.go
+++ b/cdc/sink/mysql/mysql_syncpoint_store.go
@@ -25,7 +25,6 @@ import (
        gmysql "github.com/go-sql-driver/mysql"
        "github.com/pingcap/errors"
        "github.com/pingcap/log"
-       "github.com/pingcap/tidb/parser/mysql"
        "github.com/pingcap/tiflow/cdc/contextutil"
        "github.com/pingcap/tiflow/cdc/model"
        "github.com/pingcap/tiflow/pkg/config"
@@ -211,18 +210,6 @@ func (s *mysqlSyncPointStore) SinkSyncPoint(ctx context.Context,
                return cerror.WrapError(cerror.ErrMySQLTxnError, err)
        }
 
-       // we set `tidb_enable_external_ts_read` = off for each txn,
-       // otherwise TiCDC may not be able to write data to downstream TiDB when
-       // the downstream tidb_enable_external_ts_read = on.
-       if err = s.setTidbEnableExternalTsReadOff(ctx, tx); err != nil {
-               log.Info("`set tidb_enable_external_ts_read = off` fail", zap.Error(err))
-               err2 := tx.Rollback()
-               if err2 != nil {
-                       log.Error("failed to write syncpoint table", zap.Error(err))
-               }
-               return cerror.WrapError(cerror.ErrMySQLTxnError, err)
-       }
-
        row := tx.QueryRow("select @@tidb_current_ts")
        var secondaryTs string
        err = row.Scan(&secondaryTs)
@@ -291,20 +278,3 @@ func (s *mysqlSyncPointStore) Close() error {
        err := s.db.Close()
        return cerror.WrapError(cerror.ErrMySQLConnectionError, err)
 }
-
-func (s *mysqlSyncPointStore) setTidbEnableExternalTsReadOff(
-       ctx context.Context, txn *sql.Tx,
-) error {
-       // We should always try to set this variable, and ignore the error if
-       // downstream does not support this variable, it is by design.
-       query := fmt.Sprintf("SET SESSION %s = %s", "tidb_enable_external_ts_read", "off")
-       _, err := txn.ExecContext(ctx, query)
-       if err != nil {
-               if mysqlErr, ok := errors.Cause(err).(*gmysql.MySQLError); ok &&
-                       mysqlErr.Number == mysql.ErrUnknownSystemVariable {
-                       return nil
-               }
-               return err
-       }
-       return nil
-}
diff --git a/cdc/sinkv2/eventsink/txn/mysql/mysql.go b/cdc/sinkv2/eventsink/txn/mysql/mysql.go
index 0da776428..c73a354e8 100644
--- a/cdc/sinkv2/eventsink/txn/mysql/mysql.go
+++ b/cdc/sinkv2/eventsink/txn/mysql/mysql.go
@@ -536,24 +536,6 @@ func (s *mysqlBackend) execDMLWithMaxRetries(pctx context.Context, dmls *prepare
                                        start, s.changefeed, "BEGIN", dmls.rowCount, dmls.startTs)
                        }
 
-                       // we set `tidb_enable_external_ts_read` = off for each txn,
-                       // otherwise TiCDC may not be able to write data to downstream TiDB when
-                       // the downstream tidb_enable_external_ts_read = on.
-                       if err = s.setTidbEnableExternalTsReadOff(ctx, tx); err != nil {
-                               err := logDMLTxnErr(
-                                       cerror.WrapError(cerror.ErrMySQLTxnError, err),
-                                       start, s.changefeed,
-                                       fmt.Sprintf("SET SESSION %s = %s", "tidb_enable_external_ts_read",
-                                               "off"),
-                                       dmls.rowCount, dmls.startTs)
-                               if rbErr := tx.Rollback(); rbErr != nil {
-                                       if errors.Cause(rbErr) != context.Canceled {
-                                               log.Warn("failed to rollback txn", zap.Error(rbErr))
-                                       }
-                               }
-                               return 0, err
-                       }
-
                        for i, query := range dmls.sqls {
                                args := dmls.values[i]
                                log.Debug("exec row", zap.Int("workerID", s.workerID),
diff --git a/pkg/sink/mysql/db_helper.go b/pkg/sink/mysql/db_helper.go
index 9ccc70af8..2ce16ee11 100644
--- a/pkg/sink/mysql/db_helper.go
+++ b/pkg/sink/mysql/db_helper.go
@@ -160,6 +160,15 @@ func generateDSNByConfig(
        if tidbPlacementMode != "" {
                dsnCfg.Params["tidb_placement_mode"] = fmt.Sprintf(`"%s"`, tidbPlacementMode)
        }
+
+       tidbEnableExternalTSRead, err := checkTiDBVariable(ctx, testDB, "tidb_enable_external_ts_read", "OFF")
+       if err != nil {
+               return "", err
+       }
+       if tidbEnableExternalTSRead != "" {
+               // set the `tidb_enable_external_ts_read` to `OFF`, so cdc could write to the sink
+               dsnCfg.Params["tidb_enable_external_ts_read"] = fmt.Sprintf(`"%s"`, tidbEnableExternalTSRead)
+       }
        dsnClone := dsnCfg.Clone()
        dsnClone.Passwd = "******"
        log.Info("sink uri is configured", zap.String("dsn", dsnClone.FormatDSN()))
```

There are two identical functions `generateDSNByParams` in two different files: `pkg/sink/mysql/db_helper.go`, `cdc/sink/mysql/mysql_params.go`. It seems that we have to modify both to make it work :thinking:

**Pull request merged**